netutils/netinit/netinit.c

```diff
@@ -221,13 +221,22 @@
 #  define AF_INETX AF_INET6
 #endif
 
-/* While the network is up, the network monitor really does nothing.  It
- * will wait for a very long time while waiting, it can be awakened by a
- * signal indicating a change in network status.
+/* If using signal notification, raise by a the PHY, then while the network
+ * is up, the network monitor really does nothing.  It will wait for a very
+ * long time while waiting, it can be awakened by a signal indicating a
+ * change in network status.
+ *
+ * If the network monitor is used in polled mode these values are set by the
+ * CONFIG_NETINIT_LOSS_POLL_RATE and CONFIG_NETINIT_ESTABLISH_POLL_RATE
  */
 
-#define LONG_TIME_SEC    (60*60) /* One hour in seconds */
-#define SHORT_TIME_SEC   (2)     /* 2 seconds */
+#if !(defined(CONFIG_ARCH_PHY_POLLED))
+#  define LONG_TIME_SEC    (60*60) /* One hour in seconds */
```

Comment:
Why not make NETINIT_ESTABLISH_POLL_RATE/NETINIT_LOSS_POLL_RATE available when ARCH_PHY_INTERRUPT enable by removing the dependence(depends on !ARCH_PHY_INTERRUPT && ARCH_PHY_POLLED)?

Reply:
If you look ad the code this is enabled from the bottom up.  Formally by ARCH_PHY_INTERRUPT and now also by ARCH_PHY_INTERRUPT.

It is this way because I am respecting the previous functionality 100%. I find that to be respectful of all users. I explained to you before I do not have HW with a PHY that has an IRQ, I test my changes.  This approach ensures I did not break other peoples code.  If you can test it with **all** the options submit a PR and I will be happy to retest what I need working.

Reply:
I don't think the below code change will break the previous functionality:
```
config NETINIT_ESTABLISH_POLL_RATE
	int "The poll rate in seconds, to check for link establishment."
	default 2
	---help---
		The network monitor will check the PHY link state every
		NETINIT_ESTABLISH_POLL_RATE seconds for link establishment.

config NETINIT_LOSS_POLL_RATE
	int "The poll rate in seconds, to check for link loss."
	default 3600
	---help---
		The network monitor will check the PHY link state every
		NETINIT_LOSS_POLL_RATE seconds for link loss.

#  define LONG_TIME_SEC   CONFIG_NETINIT_LOSS_POLL_RATE
#  define SHORT_TIME_SEC CONFIG_NETINIT_ESTABLISH_POLL_RATE
```
It's the responsibility of maintainer to keep the code quality and avoid the code become "spanish noodles" as possible as we can.

Reply:
I understand the original code to do the right thing in using interrupts.  I will not break that, nor should the user be exposed to it.

Reply:
I read the code long time ago, but if my memory is correct, the polling logic is already supported to handle the hardware which can't generate the interrupt with the correct timing. It's very easy to support the PHY which can't generate the interrupt at all by making LONG_TIME_SEC/SHORT_TIME_SEC configurable.

Reply:
> I read the code long time ago, but if my memory is correct,

your recollection is not correct.

I ask the question on dev first - see ponymail for Greg's response and his concern about management while in link active. 

It does not support polling.

All of this has been discussed. 

See the title block:
```
Not all boards have in interrupt line from the phy to the Soc. This allows the phy to be polled for link status.
This may not work on all MAC/PHY combination that have mutually exclusive link management and operating
modes. ***********************The STM32F7 and LAN8742AI do not have such a limitation.********

To use this option the arch must supply CONFIG_ARCH_PHY_POLLED and not CONFIG_ARCH_PHY_INTERRUPT.
```
ARCH_PHY_POLLED - is the enabler set by the HW knowing it can deal with polling. 

Without being intimately familiar with the code and all the issues that have been addressed you are not adding value here.

Reply:
Why do you think that I am not familiar with the net monitor code? I fixed a bug in it several months ago:
https://github.com/apache/incubator-nuttx-apps/commit/d96bc5a4fbb90754f5afececbdc97f3d5395076e
And create an issue suggest to decouple the monitor from ethernet hardware:
https://github.com/apache/incubator-nuttx-apps/commit/d96bc5a4fbb90754f5afececbdc97f3d5395076e

Let's come back to your change, here is the original code snippet:
···
static int netinit_monitor(void)
{
  /* Now loop, waiting for changes in link status */

  for (; ; )
    {
      /* Check for link up or down */

      if ((ifr.ifr_mii_val_out & MII_MSR_LINKSTATUS) != 0)
        {
          /* Link up... does the drive think that the link is up? */

          if (!devup)
            {
              /* No... We just transitioned from link down to link up.
               * Bring the link up.
               */

              /* And wait for a short delay.  We will want to recheck the
               * link status again soon.
               */

              reltime.tv_sec  = SHORT_TIME_SEC;
              reltime.tv_nsec = 0;
            }
          else
            {
              /* The link is still up.  Take a long, well-deserved rest */

              reltime.tv_sec  = LONG_TIME_SEC;
              reltime.tv_nsec = 0;
            }
        }
      else
        {
          /* Link down... Was the driver link state already down? */

          if (devup)
            {
              /* No... we just transitioned from link up to link down.  Take
               * the link down.
               */
                }
            }

          /* In either case, wait for the short, configurable delay */

          reltime.tv_sec  = CONFIG_NETINIT_RETRYMSEC / 1000;
          reltime.tv_nsec = (CONFIG_NETINIT_RETRYMSEC % 1000) * 1000000;
        }

      /* Now wait for either the semaphore to be posted for a timed-out to
       * occur.
       */

      sched_lock();
      DEBUGVERIFY(clock_gettime(CLOCK_REALTIME, &abstime));

      abstime.tv_sec  += reltime.tv_sec;
      abstime.tv_nsec += reltime.tv_nsec;
      if (abstime.tv_nsec >= 1000000000)
        {
          abstime.tv_sec++;
          abstime.tv_nsec -= 1000000000;
        }

      sem_timedwait(&g_notify_sem, &abstime);
      sched_unlock();
    }
}
···
The keypoint for polling is sem_timedwait which already exist for a long time. What your change is:
1. Skip the signal related setting if ARCH_PHY_POLLED
2. Make SHORT_TIME_SEC/LONG_TIME_SEC configurable
All change looks good, what I suggest is making Kconfig available in both(interrupt/polled) cases.

Reply:
AGAIN **_I do not have HW to test your use case_**. If you test it. If it woks, do a PR again this branch.

Reply:
Here is my suggestion:
![image](https://user-images.githubusercontent.com/18066964/94987695-c1d50e00-059a-11eb-9b96-7f94cedb0a82.png)
![image](https://user-images.githubusercontent.com/18066964/94987779-8f77e080-059b-11eb-904c-4b6205f73f0b.png)
There isn't any logic change and the timeout also same as before. If you can't guarantee the correctness, which mean you don't have confidence with your own change.